library/ssl_misc.h

```diff
@@ -1995,4 +1997,57 @@ int mbedtls_ecp_tls_13_write_group( const mbedtls_ecp_group *grp,
 #endif /* MBEDTLS_ECP_C */
 #endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */
 
+/**
+ * To let TLS1.2 and TLS1.3 work together, we divide the task into two steps.
+ * Below inline functions are required by first step.
+ * 1. Configured as TLS1.2 or TLS1.3. Identify the protocol version based on configuration field.
+ * 2. Configured as TLS1.2 and TLS1.3. Identify the protocal version based on runtime parameter.
+ */
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+static inline int is_configured_as_tls1_3(const mbedtls_ssl_config *conf)
+{
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2)
+    if(conf->max_major_ver==MBEDTLS_SSL_MAJOR_VERSION_3
+        && conf->min_major_ver == MBEDTLS_SSL_MAJOR_VERSION_3
+        && conf->max_minor_ver == MBEDTLS_SSL_MINOR_VERSION_4
+        && conf->min_minor_ver == MBEDTLS_SSL_MINOR_VERSION_4)
+        return( 1 );
+    return( 0 );
+#else /* defined(MBEDTLS_SSL_PROTO_TLS1_2) */
+    (( void )conf);
+    return( 1 );
+#endif /* defined(MBEDTLS_SSL_PROTO_TLS1_2) */
+}
+#endif /* defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) */
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2)
+static inline int is_configured_as_tls1_2_or_earlier(const mbedtls_ssl_config *conf)
+{
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+    if(conf->max_major_ver<=MBEDTLS_SSL_MAX_MAJOR_VERSION
+        && conf->min_major_ver >= MBEDTLS_SSL_MIN_MAJOR_VERSION
+        && conf->max_minor_ver <= MBEDTLS_SSL_MINOR_VERSION_3
+        && conf->min_minor_ver >= MBEDTLS_SSL_MIN_MINOR_VERSION)
+        return( 1 );
+    return( 0 );
+#else /* defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) */
+    (( void )conf);
+    return( 1 );
+#endif /* defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) */
+}
+#endif /* defined(MBEDTLS_SSL_PROTO_TLS1_2) */
+
```

Comment:
We need a sanity check in `mbedtls_ssl_setup()` that the config is indeed 1.2-only or 1.3-only.

In anticipation of further config sanity checks, I'd suggest you add a static function `ssl_config_check()` which is called at the very beginning of `mbedtls_ssl_setup()`. For now, this function just checks that `ssl_conf_is_tls13_only() || ssl_conf_is_tls12_only()` -- but more might be added at a later point.

Reply:
I create another PR(#337) for that . 
This PR depend on that now. And #322 also depend on that